File: python-project-euler-100/p014.py
from Solution import Solution

class P014(Solution):
    mem_cache = {1: 0}

    def collatz(self, number):
        if number in self.mem_cache:
            return self.mem_cache[number]

        if number % 2 == 0:
            data = self.collatz((number // 2))
            self.mem_cache[number] = data + 1
            return (1 + data)

        if number % 2 == 1:
            data = self.collatz((3 * number + 1))
            self.mem_cache[number] = data + 1
            return (1 + data)

    def solve(self):
        self.problem_number = 14
        last = 1000000
        max_number = 1
        max_length = self.mem_cache[max_number];

        for i in range(2, last):
            sequence_length = self.collatz(i)
            if sequence_length > max_length:
                max_length = sequence_length
                max_number = i

        return max_number

def main():
	P014().run()

if __name__ == "__main__":
	main()

File: go-project-euler-100/p014.go
package main

var p014Cache = map[int]int{1: 0}

func p014Collatz(n int) int {
	if v, ok := p014Cache[n]; ok {
		return v
	}
	var data int
	if n%2 == 0 {
		data = p014Collatz(n/2) + 1
	} else {
		data = p014Collatz(3*n+1) + 1
	}
	p014Cache[n] = data
	return data
}

/*P014Solve solution for p014*/
func P014Solve() int {
	maxNumber := 1
	maxLength := 0
	last := 1000000

	for i := 2; i < last; i++ {
		seqLen := p014Collatz(i)
		if seqLen > maxLength {
			maxLength = seqLen
			maxNumber = i
		}
	}

	return maxNumber
}
